crates/wasmparser/src/validator/operators.rs

```diff
@@ -1416,6 +1428,65 @@ where
         }
     }
 
+    fn cont_type_at(&self, key: TypeKey) -> Result<&ContType> {
+        let sub_ty = match key {
+            Either::A(at) => self.sub_type_at(at)?,
+            Either::B(id) => match self.resources.sub_type_at_id(id) {
+                None => bail!(self.offset, "unknown type {}", TypeIdentifier::index(&id)),
+                Some(sub_ty) => sub_ty,
+            },
+        };
+        if let CompositeInnerType::Cont(cont_ty) = &sub_ty.composite_type.inner {
+            if self.inner.shared && !sub_ty.composite_type.shared {
+                bail!(
+                    self.offset,
+                    "shared continuations cannot access unshared continuations",
+                );
+            }
+            Ok(cont_ty)
+        } else {
+            bail!(
+                self.offset,
+                "non-continuation type {}",
+                match key {
+                    Either::A(at) => at as usize,
+                    Either::B(id) => TypeIdentifier::index(&id),
+                }
+            )
+        }
+    }
```

Comment:
I might recommend splitting this into `cont_type_at` which takes a `u32` and `const_type_at_id` which takes a `CoreTypeId`. I think that might be a bit easier to work with instead of working with `Either` perhaps? Functions like `func_type_of_cont_type_at` below could take `&ContType` instead of `TypeKey` in that case to let the caller call the appropriate `const_type_at*` function and then pass the result into later parts of validation.